src/sentry/api/endpoints/organization_teams.py

```diff
@@ -142,56 +145,88 @@ def post(self, request: Request, organization, **kwargs) -> Response:
+                    {"detail": "You do not have permission to join a new team as a Team Admin"},
                 )
-            else:
-                team_created.send_robust(
-                    organization=organization, user=request.user, team=team, sender=self.__class__
+        try:
+            with transaction.atomic():
+                team = Team.objects.create(
+                    name=result.get("name") or result["slug"],
+                    slug=result.get("slug"),
+                    idp_provisioned=result.get("idp_provisioned", False),
+                    organization=organization,
                 )
-            if self.should_add_creator_to_team(request):
-                try:
+        except (IntegrityError, MaxSnowflakeRetryError):
+            raise ConflictError(
+                {
+                    "non_field_errors": [CONFLICTING_SLUG_ERROR],
+                    "detail": CONFLICTING_SLUG_ERROR,
+                }
+            )
+        else:
+            team_created.send_robust(
+                organization=organization,
+                user=request.user,
+                team=team,
+                sender=self.__class__,
+            )
+
+        if self.should_add_creator_to_team(request):
+            try:
+                with transaction.atomic():
                     member = OrganizationMember.objects.get(
                         user=request.user, organization=organization
                     )
-                except OrganizationMember.DoesNotExist:
-                    pass
-                else:
-                    OrganizationMemberTeam.objects.create(team=team, organizationmember=member)
-
-            self.create_audit_entry(
-                request=request,
-                organization=organization,
-                target_object=team.id,
-                event=audit_log.get_event_id("TEAM_ADD"),
-                data=team.get_audit_log_data(),
-            )
-            return Response(
-                serialize(team, request.user, self.team_serializer_for_post()),
-                status=201,
-            )
-        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+                    OrganizationMemberTeam.objects.create(
+                        team=team,
+                        organizationmember=member,
+                        role="admin" if set_team_admin else None,
+                    )
+            except OrganizationMember.DoesNotExist:
+                if set_team_admin:
+                    # delete team if we can't add the user as a team admin
+                    team.delete()
+                    raise PermissionDenied(
+                        detail="You must be a member of the organization to join a new team as a Team Admin"
+                    )
+            # catch generic exception to ensure the team gets deleted if we can't add the user as
+            # a Team Admin
+            except Exception as e:
```

Comment:
Use a `finally` block to delete the team. Catching a generic Exception and raising a ValidationError is risky as critical errors that we don't want to be caught, would be caught and swallowed by the ValidationError.

Reply:
You should consolidate this with other team cleanups in the try/catch block.

Reply:
Changed to using `finally` w/ boolean flag `cleanup_needed`